### PR TITLE
Disable auto-configuration of Hazelcast IMDG when Hazelcast Jet is present

### PR DESCRIPTION
**Pull request by gurbuzali**
Spring Boot auto-configures Hazelcast IMDG when it finds relevant configuration file on the classpath or root directory. Hazelcast Jet wraps IMDG and uses the very same configuration files(in addition to Hazelcast Jet specific configuration file) to configure and create Hazelcast Jet instances. This PR disables Hazelcast IMDG auto-configuration if Jet presents by checking a certain file (`hazelcast-jet-default.yaml`) is on the classpath. 

See discussion on this PR https://github.com/spring-projects/spring-boot/pull/19483

fixes https://github.com/hazelcast/hazelcast-jet/issues/1835

See [Hazelcast Jet Starter Spring Boot Starter](https://github.com/hazelcast/hazelcast-jet-contrib/tree/master/hazelcast-jet-spring-boot-starter)

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastConfigResourceCondition.java [hunk 1]
```diff
@@ -43,6 +46,16 @@ protected HazelcastConfigResourceCondition(String configSystemProperty, String..
 		this.configSystemProperty = configSystemProperty;
 	}
 
+	@Override
+	public ConditionOutcome getMatchOutcome(ConditionContext context, AnnotatedTypeMetadata metadata) {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastConfigResourceCondition.java]: This code doesn't match with the javadoc or name of `HazelcastConfigResourceCondition` so it shouldn't be located there.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastAutoConfigurationJetTests.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Test for {@link HazelcastAutoConfiguration} when Hazelcast Jet is present. Spring Boot
+ * should not auto-create Hazelcast IMDG instances when Hazelcast Jet is present.
+ *
+ * @author Ali Gurbuz
+ */
+@ClassPathOverrides("com.hazelcast.jet:hazelcast-jet:4.0")
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastAutoConfigurationJetTests.java]: I prefer no to refer to a concept if we can avoid doing so. This test could be put in the main test with a special `ClassLoader` that returns the resource.

**Comment by snicoll** [comment]
@gurbuzali are you aware a colleague of yours has already submitted an attempt? see #19483. Ignoring the class check, I think it's more aligned with what we need to do there (besides testing as I've mentioned in a comment).

**Comment by gurbuzali** [comment]
Hi @snicoll, I've tried to address your review comments, please take a look.

**Comment by snicoll** [comment]
Thanks for making your first contribution to Spring Boot.

**Pull request closed without merging**